services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/CompletedFileUpload.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.transfer.s3;
+
+import java.util.Objects;
+import software.amazon.awssdk.annotations.SdkPreviewApi;
+import software.amazon.awssdk.annotations.SdkPublicApi;
+import software.amazon.awssdk.services.s3.S3AsyncClient;
+import software.amazon.awssdk.services.s3.model.PutObjectResponse;
+import software.amazon.awssdk.utils.ToString;
+import software.amazon.awssdk.utils.Validate;
+
+/**
+ * Represents a completed upload transfer to Amazon S3. It can be used to track
+ * the underlying {@link PutObjectResponse}
+ *
+ * @see S3TransferManager#uploadFile(UploadFileRequest)
+ */
+@SdkPublicApi
+@SdkPreviewApi
+public final class CompletedFileUpload implements CompletedObjectTransfer {
+    private final PutObjectResponse response;
+
+    private CompletedFileUpload(DefaultBuilder builder) {
+        this.response = Validate.paramNotNull(builder.response, "response");
+    }
+    
+    public PutObjectResponse response() {
+        return response;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        CompletedFileUpload that = (CompletedFileUpload) o;
+
+        return Objects.equals(response, that.response);
+    }
+
+    @Override
+    public int hashCode() {
+        return response != null ? response.hashCode() : 0;
```

Comment:
Shouldn't need null checks for non-null fields. Worth checking all objects for this, sorry. Hackathon project: codegen these stupid objects.

Reply:
A tool that will stop working for future Java versions :) (the bear has been poked)

Reply:
A tool that's a build-time dependency only, and for which we'll typically have years to react to any such discontinuation (which I'm convinced won't happen without an official replacement)!